### PR TITLE
Sidenotes

### PR DESCRIPTION
**Pull request by jimrandomh**
Adds support for sidenotes, which are a way of displaying footnotes in the right margin of the page. They are automatically extracted from the post; footnotes on existing posts are displayed this way. When sidenotes would overlap, they instead push each other down (in practice, into free space).

This also makes side-comments and inline reactions work similarly. Before inline reactions were represented with an underline inside the post text; now they're represented with an icon in the right margin.

Screenshot 2024-08-20 at 11 30 58

On mobile, side-comments and inline reactions are represented by a small handle in the right margin, like this:

Screenshot 2024-08-20 at 11 41 28

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1208106054788926) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by jimrandomh** [comment]
Typecheck failure in `postResolvers.ts` is a (conceptual) merge conflict with the denormalized-contents branch, with the symptom that auto-sizing the right column based on whether there are sidenotes or not may not work. Looking into it.

**Comment by Raemon** [comment]
It's a little sad that when I mouse over an expandable sidenote, then try to mouseover the one below it, my cursor ends up positioned so it ends up in a random place instead.

I'm not sure what to do instead, mostly just noting it as a thing to think about for the future.

https://github.com/user-attachments/assets/6896d59b-3db0-4324-b327-ddcf6a0c44e8

**Comment by Raemon** [comment]
Another fiddly UI think that seems hard to fix, but sad, is when things end up right next to each other:
image

**Comment by Will-Howard** [comment]
It's getting the numbering wrong in this case, it goes through 1,2,3 in the sidenotes and then starts again from 1 (the text for footnote 4 is correct, but the number is wrong):

Screenshot 2024-08-22 at 14 18 46

This is the post if you want to try copying the contents: https://forum.effectivealtruism.org/posts/vbhoFsyQmrntru6Kw/do-brains-contain-many-conscious-subsystems-if-so-should-we

File: packages/lesswrong/components/linkPreview/FootnotePreview.tsx [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
+import React, { useCallback, useContext, useState } from 'react';
 import Card from '@material-ui/core/Card';
 import { Components, registerComponent } from '../../lib/vulcan-lib';
 import { useHover } from '../common/withHover';
 import { EXPAND_FOOTNOTES_EVENT } from '../posts/PostsPage/CollapsedFootnotes';
-import { hasCollapsedFootnotes } from '@/lib/betas';
+import { hasCollapsedFootnotes, hasSidenotes } from '@/lib/betas';
+import classNames from 'classnames';
+import { SideItemContentContext } from '../contents/SideItems';
+import { truncate } from '@/lib/editor/ellipsize';
```
Comment (Will-Howard) [inline packages/lesswrong/components/linkPreview/FootnotePreview.tsx:9]: unused

File: packages/lesswrong/server/resolvers/postResolvers.ts [hunk 2]
```diff
@@ -16,7 +16,7 @@ import { postIsCriticism } from '../languageModels/autoTagCallbacks';
 import { createPaginatedResolver } from './paginatedResolver';
 import { getDefaultPostLocationFields, getDialogueResponseIds, getDialogueMessageTimestamps, getPostHTML } from "../posts/utils";
 import { buildRevision } from '../editor/make_editable_callbacks';
-import { cheerioParse } from '../utils/htmlUtil';
+import { cheerioParse, htmlContainsFootnotes } from '../utils/htmlUtil';
```
Comment (Will-Howard) [inline packages/lesswrong/server/resolvers/postResolvers.ts:19]: unused

**Review by Will-Howard — COMMENTED** [review]
Sorry if the unused import comments are annoying, I'm not sure how done this is supposed to be.

It does change the column spacing for the EAF version of the post page, so we would like that to match what it was before:
Screenshot 2024-08-22 at 17 46 56
Screenshot 2024-08-22 at 17 47 15

File: packages/lesswrong/components/contents/SideItemLine.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    width: 4,
+    height: 24,
+    position: "relative",
+    left: 9,
+    borderTopLeftRadius: 4,
+    borderBottomLeftRadius: 4,
+  },
+})
+
+const SideItemLine = ({colorClass, classes}: {
```
[the remark this answers is not in the document]
Comment (jimrandomh) [reply packages/lesswrong/components/contents/SideItemLine.tsx:22]: "Line" refers to the visual style of the indicator (it's like a vertical line)

**Comment by jimrandomh** [comment]
Column spacing was changed on EA forum due to some styles missing their forum-gating, which should be fixed now. The incorrect footnote numbering on an EA Forum post shown by Will was caused by an interaction with the collapsing-footnotes feature, which isn't used on LW, but should be fixed now nevertheless.

File: packages/lesswrong/components/comments/SideCommentIcon.tsx [hunk 4]
```diff
@@ -93,11 +101,63 @@ const BadgeWrapper = ({commentCount, classes, children}: {
 const SideCommentIcon = ({commentIds, post, classes}: {
   commentIds: string[]
   post: PostsList
-  classes: ClassesType
+  classes: ClassesType<typeof styles>
+}) => {
+  const isMobile = useIsMobile();
```
[the remark this answers is not in the document]
Comment (jimrandomh) [reply packages/lesswrong/components/comments/SideCommentIcon.tsx:106]: `useIsMobile` is just a no-complications alias for `!useIsAboveBreakpoint('sm')`. It's indeed not SSR-safe, but fortunately this component doesn't ever run in an SSR context.

File: packages/lesswrong/components/comments/SideCommentIcon.tsx [hunk 5]
```diff
@@ -93,11 +101,63 @@ const BadgeWrapper = ({commentCount, classes, children}: {
...
+  if (isMobile) {
+    return <SideCommentIconMobile commentIds={commentIds} post={post} classes={classes} />;
+  } else {
+    return <SideCommentIconDesktop commentIds={commentIds} post={post} classes={classes} />;
```
[the remark this answers is not in the document]
Comment (jimrandomh) [reply packages/lesswrong/components/comments/SideCommentIcon.tsx:110]: Not great, but this was already the side-comment behavior before. You can horiz-scroll to see it.

**Comment by jimrandomh** [comment]
Unit/integration test failure is because coveralls.io is down (test log shows a 405 from coveralls.io, and their web site is a big down for maintenance message right now.)

File: packages/lesswrong/components/comments/SideCommentIcon.tsx [hunk 6]
```diff
@@ -230,8 +309,8 @@ const SideCommentSingle = ({commentId, post, dontTruncateRoot=false, classes}: {
   //   eslint-disable-next-line react-hooks/exhaustive-deps
   }, [commentId, hoveredBlockquoteId, rootDivRef.current]);
   
-  if (loading) return <Loading/>
-  if (!comment) return null;
+  // if (loading) return <Loading/>
```
[the remark this answers is not in the document]
Comment (jimrandomh) [reply packages/lesswrong/components/comments/SideCommentIcon.tsx]: To delete, actually; this stopped being correct when we added the optimistic-loading stuff earlier in the function.

**Pull request merged**